File: app/src/test/java/org/ole/planet/myplanet/services/sync/AdaptiveBatchProcessorTest.kt
package org.ole.planet.myplanet.services.sync

import org.junit.Assert.assertEquals
import org.junit.Test

class AdaptiveBatchProcessorTest {
    @Test
    fun `starts at initial size`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100)
        assertEquals(100, processor.currentSize)
    }

    @Test
    fun `initial size is clamped to min`() {
        val processor = AdaptiveBatchProcessor(initialSize = 5, minSize = 10)
        assertEquals(10, processor.currentSize)
    }

    @Test
    fun `failure halves the batch size`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100)
        processor.recordFailure()
        assertEquals(50, processor.currentSize)
    }

    @Test
    fun `repeated failures never drop below min size`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100, minSize = 10)
        repeat(10) { processor.recordFailure() }
        assertEquals(10, processor.currentSize)
    }

    @Test
    fun `slow response halves the batch size`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100, slowThresholdMs = 8_000)
        processor.recordSuccess(durationMs = 9_000)
        assertEquals(50, processor.currentSize)
    }

    @Test
    fun `fast response grows the batch size`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100, maxSize = 200, fastThresholdMs = 2_000)
        processor.recordSuccess(durationMs = 500)
        assertEquals(126, processor.currentSize)
    }

    @Test
    fun `growth never exceeds max size`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100)
        repeat(10) { processor.recordSuccess(durationMs = 100) }
        assertEquals(100, processor.currentSize)
    }

    @Test
    fun `moderate response keeps size unchanged`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100, slowThresholdMs = 8_000, fastThresholdMs = 2_000)
        processor.recordSuccess(durationMs = 5_000)
        assertEquals(100, processor.currentSize)
    }

    @Test
    fun `recovers toward max after a slow spell`() {
        val processor = AdaptiveBatchProcessor(initialSize = 100)
        processor.recordFailure()
        processor.recordFailure()
        assertEquals(25, processor.currentSize)
        repeat(20) { processor.recordSuccess(durationMs = 100) }
        assertEquals(100, processor.currentSize)
    }
}

File: flutter/test/core/sync/adaptive_batch_processor_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:myplanet/core/sync/adaptive_batch_processor.dart';

void main() {
  group('AdaptiveBatchProcessor', () {
    test('clamps the initial size into range', () {
      expect(
        AdaptiveBatchProcessor(
          initialSize: 1,
          minSize: 10,
          maxSize: 100,
        ).currentSize,
        10,
      );
      expect(
        AdaptiveBatchProcessor(
          initialSize: 500,
          minSize: 10,
          maxSize: 100,
        ).currentSize,
        100,
      );
    });

    test('halves the batch on a slow response, down to minSize', () {
      final processor = AdaptiveBatchProcessor(
        initialSize: 100,
        minSize: 10,
        maxSize: 100,
      );
      processor.recordSuccess(9000);
      expect(processor.currentSize, 50);
      processor.recordSuccess(9000);
      expect(processor.currentSize, 25);
      for (var i = 0; i < 10; i++) {
        processor.recordSuccess(9000);
      }
      expect(processor.currentSize, 10);
    });

    test('grows the batch on a fast response, up to maxSize', () {
      final processor = AdaptiveBatchProcessor(
        initialSize: 20,
        minSize: 10,
        maxSize: 100,
      );
      processor.recordSuccess(500);
      expect(processor.currentSize, 26);
      for (var i = 0; i < 20; i++) {
        processor.recordSuccess(500);
      }
      expect(processor.currentSize, 100);
    });

    test('holds steady between the thresholds', () {
      final processor = AdaptiveBatchProcessor(
        initialSize: 50,
        minSize: 10,
        maxSize: 100,
      );
      processor.recordSuccess(5000);
      expect(processor.currentSize, 50);
    });

    test('halves the batch on failure', () {
      final processor = AdaptiveBatchProcessor(
        initialSize: 80,
        minSize: 10,
        maxSize: 100,
      );
      processor.recordFailure();
      expect(processor.currentSize, 40);
    });

    test('defaults maxSize to the initial size', () {
      final processor = AdaptiveBatchProcessor(initialSize: 40);
      processor.recordSuccess(100);
      expect(processor.currentSize, 40);
    });
  });
}
